File: datasets/Part 3 - Classification/Section 15 - K-Nearest Neighbors (K-NN)/knn.R
# K - Nearest Neighbors (K-NN)

# =======================================================================================================
# IDEA
#
# El algoritmo parte de tener una serie de datos categorizados. Al aparecer un nuevo dato, el algoritmo
# a que conjunto pertenece.
#
# PASOS:
# 1.- Elegir el numero "k" de vecinos del nuevo dato se tomaran en cuenta para la clasificacion. Se recomienda
#     un numero impar, para evitar empates. K = 5 es muy usado.
# 2.- Se toman los "k" veninos mas cercanos del nuevo dato; utilizando distancias euclidianas la mayoria
#     de las ocasiones, aunque tambien se puede utilizar la distancia Manhattan del valor absoluto, la
#     distancia Minkowski, metrica del infinito, etc.
# 3.- Ver a que categoria pertenecen los "k" vecinos y contar cuantos pertenecen a cada categpria
# 4.- Asignar el nuevo dato a la categoria con mas vecinos.
#
# =======================================================================================================
# 
# La frontera entre las clases practicamente nunca es lineal.
#
# =======================================================================================================

################################################
###          IMPORTAR EL DATA SET            ###
################################################

setwd("~/Documentos/Udemy/machinelearning-az/datasets/Part 3 - Classification/Section 15 - K-Nearest Neighbors (K-NN)")
dataset = read.csv('Social_Network_Ads.csv')
dataset = dataset[, 3:5]

#################################################################################
### Dividir el data set en conjunto de entrenamiento y conjunto de testing    ###
#################################################################################

# install.packages("caTools")
library(caTools)
set.seed(123)
split = sample.split(dataset$Purchased, SplitRatio = 0.75)
training_set = subset(dataset, split == TRUE)
testing_set = subset(dataset, split == FALSE)

################################################
#            Escalado de variables             #
################################################

training_set[,1:2] = scale(training_set[,1:2])
testing_set[,1:2] = scale(testing_set[,1:2])

#############################################################################################################
##   Ajustar el modelo de KNN con todo el dataset de Entrenamiento y hacer la REDICCION con el de Testing   #
#############################################################################################################

# Cargamos la libreria "class" que contiene un conjunto de funciones para procesos de clasificacion.
library(class)
# La funcion devuelve la prediccion, recibe como parametros a lo conjuntod de Entrenamiento
# y Testing. ¡OJO! "train" hace referencia a la matriz de caracteristicas de Entrenamiento (variables independientes)
# mientras que "cl", hace referencia a la clasificaciones verdaderas del conjunto de Entrenamiento.

# El parametro "k", hace referencia al numero de "k" vecinos para clasificar
# a los nuevos datos.
y_pred = knn(train = training_set[,-3],
             test = testing_set[,-3],
             cl = training_set[,3],
             k = 5)

# La prediccion es un vector con con cada una de las clases elegidas (0 o 1), ya no es la probabilidad.

################################################
#        EVALUACION DEL RENDIMIENTO            #
################################################

# Crear la matriz de confusión
cm = table(testing_set[, 3], y_pred)

# INTERPRETACION: Las columnas representan el dato real, mientras que las fila la prediccion.
# La diagonal de izq. a der. representa aquellos casos en los que lagoritmo acerto, mientras
# que la diagonal que va de der. a izq. representa aquellos caso en los que el algoritmo fallo.
# VN  FN
# FP  VP

################################################
#        VISUALIZACION DE RESULTADOS           #
################################################

# Visualización del conjunto de entrenamiento
#install.packages("ElemStatLearn")
library(ElemStatLearn)
set = training_set
X1 = seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.01)
X2 = seq(min(set[, 2]) - 1, max(set[, 2]) + 1, by = 0.01)
grid_set = expand.grid(X1, X2)
colnames(grid_set) = c('Age', 'EstimatedSalary')
y_grid = knn(train = training_set[,-3],
             test = grid_set,
             cl = training_set[,3],
             k = 5)
plot(set[, -3],
     main = 'K-NN (Conjunto de Entrenamiento)',
     xlab = 'Edad', ylab = 'Sueldo Estimado',
     xlim = range(X1), ylim = range(X2))
contour(X1, X2, matrix(as.numeric(y_grid), length(X1), length(X2)), add = TRUE)
points(grid_set, pch = '.', col = ifelse(y_grid == 1, 'springgreen3', 'tomato'))
points(set, pch = 21, bg = ifelse(set[, 3] == 1, 'green4', 'red3'))


# Visualización del conjunto de testing
set = testing_set
X1 = seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.01)
X2 = seq(min(set[, 2]) - 1, max(set[, 2]) + 1, by = 0.01)
grid_set = expand.grid(X1, X2)
colnames(grid_set) = c('Age', 'EstimatedSalary')
y_grid = knn(train = training_set[,-3],
             test = grid_set,
             cl = training_set[,3],
             k = 5)
plot(set[, -3],
     main = 'K-NN (Conjunto de Testing)',
     xlab = 'Edad', ylab = 'Sueldo Estimado',
     xlim = range(X1), ylim = range(X2))
contour(X1, X2, matrix(as.numeric(y_grid), length(X1), length(X2)), add = TRUE)
points(grid_set, pch = '.', col = ifelse(y_grid == 1, 'springgreen3', 'tomato'))
points(set, pch = 21, bg = ifelse(set[, 3] == 1, 'green4', 'red3'))

File: datasets/Part 3 - Classification/Section 15 - K-Nearest Neighbors (K-NN)/knn.py
#!/home/kevinml/anaconda3/bin/python3.7
# -*- coding: utf-8 -*-
"""
Created on Thu Mar 14 18:20:58 2019

@author: juangabriel
"""

# K - Nearest Neighbors (K-NN)

# =======================================================================================================
# IDEA
#
# El algoritmo parte de tener una serie de datos categorizados. Al aparecer un nuevo dato, el algoritmo
# a que conjunto pertenece.
#
# PASOS:
# 1.- Elegir el numero "k" de vecinos del nuevo dato se tomaran en cuenta para la clasificacion. Se recomienda
#     un numero impar, para evitar empates. K = 5 es muy usado.
# 2.- Se toman los "k" veninos mas cercanos del nuevo dato; utilizando distancias euclidianas la mayoria
#     de las ocasiones, aunque tambien se puede utilizar la distancia Manhattan del valor absoluto, la
#     distancia Minkowski, metrica del infinito, etc.
# 3.- Ver a que categoria pertenecen los "k" vecinos y contar cuantos pertenecen a cada categpria
# 4.- Asignar el nuevo dato a la categoria con mas vecinos.
#
# =======================================================================================================
# 
# La frontera entre las clases practicamente nunca es lineal.
#
# =======================================================================================================

# Cómo importar las librerías
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

################################################
###          IMPORTAR EL DATA SET            ###
################################################

dataset = pd.read_csv('Social_Network_Ads.csv')

X = dataset.iloc[:, [2, 3]].values
y = dataset.iloc[:, 4].values


#################################################################################
### Dividir el data set en conjunto de entrenamiento y conjunto de testing    ###
#################################################################################

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.25, random_state=0)


################################################
#            Escalado de variables             #
################################################

from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)

# En este caso no hace falta escalar la variable dependiente porque por si misma
# ya es una categoria (0 o 1).

######################################################
##   Ajustar el modelo de KNN con todo el dataset    #
######################################################

# Importamos la libreria para eobtener el modelo de KNN
# https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html
from sklearn.neighbors import KNeighborsClassifier
# El parametro "n_neighbors", hace referencia al numero de "k" vecinos para clasificar
# a los nuevos datos.
# En el parametro "metric", el valor "minkowski" es una familia de distancias;
# cuando p = 1 sale la distancia Manhattan y con p = 2 sale la distancia L" o euclidiana.
classifier = KNeighborsClassifier(n_neighbors=5, metric="minkowski", p=2)
classifier.fit(X_train, y_train)

################################################
#                PREDICCION                    #
################################################

# Predicción de los resultados con el Conjunto de Testing
y_pred = classifier.predict(X_test)

################################################
#        EVALUACION DEL RENDIMIENTO            #
################################################

# Elaborar una matriz de confusión
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)

# INTERPRETACION: Las filas representan el dato real, mientras que las columnas la prediccion.
# La diagonal de izq. a der. representa aquellos casos en los que lagoritmo acerto, mientras
# que la diagonal que va de der. a izq. representa aquellos caso en los que el algoritmo fallo.
# VN  FP
# FN  VP

################################################
#        VISUALIZACION DE RESULTADOS           #
################################################

# Representación gráfica de los resultados del algoritmo en el Conjunto de Entrenamiento
from matplotlib.colors import ListedColormap
X_set, y_set = X_train, y_train
X1, X2 = np.meshgrid(np.arange(start=X_set[:, 0].min() - 1, stop=X_set[:, 0].max() + 1, step=0.01),
                     np.arange(start=X_set[:, 1].min() - 1, stop=X_set[:, 1].max() + 1, step=0.01))
plt.contourf(X1, X2, classifier.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),
             alpha=0.75, cmap=ListedColormap(('red', 'green')))
plt.xlim(X1.min(), X1.max())
plt.ylim(X2.min(), X2.max())
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(X_set[y_set == j, 0], X_set[y_set == j, 1],
                c=ListedColormap(('red', 'green'))(i), label=j)
plt.title('K-NN (Conjunto de Entrenamiento)')
plt.xlabel('Edad')
plt.ylabel('Sueldo Estimado')
plt.legend()
plt.show()


# Representación gráfica de los resultados del algoritmo en el Conjunto de Testing
X_set, y_set = X_test, y_test
X1, X2 = np.meshgrid(np.arange(start=X_set[:, 0].min() - 1, stop=X_set[:, 0].max() + 1, step=0.01),
                     np.arange(start=X_set[:, 1].min() - 1, stop=X_set[:, 1].max() + 1, step=0.01))
plt.contourf(X1, X2, classifier.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),
             alpha=0.75, cmap=ListedColormap(('red', 'green')))
plt.xlim(X1.min(), X1.max())
plt.ylim(X2.min(), X2.max())
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(X_set[y_set == j, 0], X_set[y_set == j, 1],
                c=ListedColormap(('red', 'green'))(i), label=j)
plt.title('K-NN (Conjunto de Test)')
plt.xlabel('Edad')
plt.ylabel('Sueldo Estimado')
plt.legend()
plt.show()
